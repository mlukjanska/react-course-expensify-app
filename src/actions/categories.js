import database from '../firebase/firebase';

//ADD_CATEGORY
export const addCategory = (category) => ({
    type: 'ADD_CATEGORY',
    category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
          value = '', 
          label = ''
      } = categoryData;
      const category = { value, label };

      return database.ref(`users/${uid}/categories`).push(category).then((ref) => {
          dispatch(addCategory({
              id: ref.key,
              ... category
          }));
      });
  };
};

//EDIT_CATEGORY
export const editCategory = (id, updates) => ({
    type: 'EDIT_CATEGORY',
    id, 
    updates
});

export const startEditCategory = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/categories/${id}`).update(updates).then(() => {
            dispatch(editCategory(id, updates));
        });
    };
};

// SET_CATEGORIES
export const setCategories = (categories) => ({
    type: 'SET_CATEGORIES',
    categories
});

export const startSetCategories = () => {
    return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`)
        .once('value')
        .then((snapshot) => {
            const categories = [];
            snapshot.forEach((childSnapshot) => {
                categories.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setCategories(categories));
      });
    };
};