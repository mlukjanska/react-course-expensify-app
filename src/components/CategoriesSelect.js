import React from 'react';
import Select from 'react-select';

export default class CategoriesSelect extends React.Component {
    onCategoryChange = (category) => {
        this.props.onCategoryChange( category ? category.id : '' );
    };         
    render () {
        return (
            <div>
                {
                    this.props.categories.length === 0 ? (
                        <p>No categories</p>
                    ) : ( 
                        <Select
                            className="react-select"
                            placeholder="Select category..."
                            isClearable={true}
                            onChange={this.onCategoryChange}
                            value={ this.props.categorySelected ? this.props.categories.find((category) => category.id === this.props.categorySelected) : '' }
                            options={this.props.categories}>
                        </Select>
                        )
                }

            </div>
        )
    }
}