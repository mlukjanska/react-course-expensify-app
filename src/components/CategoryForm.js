import React from 'react';

export default class CategoryForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            value: props.category ? props.category.value : '',
            label: props.category ? props.category.label : '',
            error: ''
        };
    }
    onLabelChange = (e) => {
        const label = e.target.value;
        const value = e.target.value.toLowerCase();
        this.setState(() => ({ label }));
        this.setState(() => ({ value }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.label) {
            this.setState(() => ({ error: 'Please provide category label!' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                value: this.state.value,
                label: this.state.label
            });
        }
    }       
    render () {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input 
                    type="text"
                    className="text-input"
                    placeholder="Label"
                    value={this.state.label}                      
                    onChange={this.onLabelChange}
                />
                <div>
                    <button className="button">Add Category</button>
                </div>
            </form>
        )
    }
}