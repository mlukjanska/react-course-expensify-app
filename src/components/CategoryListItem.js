import React from 'react';
import { Link } from 'react-router-dom';

const CategoryListItem = ({id, label}) => {
    return (
          <Link className="list-item" to={`/edit-category/${id}`}>
            <div>
              <h3 className="list-item__title">{label}</h3>
            </div>    
          </Link>
      )
};

export default CategoryListItem;