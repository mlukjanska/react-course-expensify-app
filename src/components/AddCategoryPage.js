import React from 'react';
import { connect } from 'react-redux';
import CategoryForm from './CategoryForm';
import { startAddCategory } from '../actions/categories';

export class AddCategoryPage extends React.Component {
  onSubmit = (category) => {
    this.props.startAddCategory(category);
    this.props.history.push('/');
  };
  render(){
    return (
      <div>
        <h1>Add Category</h1>
        <CategoryForm 
          onSubmit={this.onSubmit}
        />
      </div>
    )
  };
}

//Abstracting away dispatch functions
const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category))
});

export default connect(undefined, mapDispatchToProps)(AddCategoryPage);