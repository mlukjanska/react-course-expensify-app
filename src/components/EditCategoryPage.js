import React from 'react';
import { connect } from 'react-redux';
import CategoryForm from './CategoryForm';
import SubCategoryForm from './SubCategoryForm';
import SubCategoryList from './SubCategoryList';
import { startEditCategory } from '../actions/categories';
import { startAddSubCategory } from '../actions/subcategories';

export class EditCategoryPage extends React.Component {
    onCategorySubmit = (category) => {
        this.props.startEditCategory(this.props.category.id, category);
        this.props.history.push('/categories');
    }; 
    onSubCategorySubmit = (subcategory) => {
        this.props.startAddSubCategory(subcategory);
    };  
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Category</h1>
                    </div>
                </div>
                <div className="content-container">  
                    <CategoryForm 
                        category={this.props.category}
                        onSubmit={this.onCategorySubmit}
                    />
                    <SubCategoryForm
                        parentkey={this.props.category.id}
                        onSubmit={this.onSubCategorySubmit}
                    />
                </div>
                <SubCategoryList
                    parentkey={this.props.category.id}
                />     
            </div>
        )
    };
}

const mapDispatchToProps = (dispatch) => ({
    startEditCategory: (id, category) => dispatch(startEditCategory(id, category)),
    startAddSubCategory: (subcategory) => dispatch(startAddSubCategory(subcategory))
});

const mapStateToProps = (state, props) => ({
    category: state.categories.find((category) => category.id === props.match.params.id),
});

export default connect(mapStateToProps,mapDispatchToProps)(EditCategoryPage);