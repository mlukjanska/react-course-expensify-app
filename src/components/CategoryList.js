import React from 'react';
import { connect } from 'react-redux';
import CategoryListItem from './CategoryListItem'

export const CategoryList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div>Categories</div>
        </div>
        <div className="list-body">
        {
            props.categories.length === 0 ? (
                <div className="list-item list-item--message">
                    <span>No categories</span>
                </div>
            ) : ( 
                    props.categories.map((category) => {
                        return <CategoryListItem key={category.id} {...category}/>;
                    })
                )
        }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        categories: state.categories
    };
}

export default connect(mapStateToProps)(CategoryList); 