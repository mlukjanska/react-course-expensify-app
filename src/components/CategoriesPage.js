import React from 'react';
import { connect } from 'react-redux';
import CategoryForm from './CategoryForm';
import CategoryList from './CategoryList';
import { startAddCategory } from '../actions/categories';

export class CategoriesPage extends React.Component {
  onSubmit = (category) => {
    this.props.startAddCategory(category);
    this.props.history.push('/');
  };
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Categories</h1>
          </div>
        </div>
        <div className="content-container">
          <CategoryForm 
            onSubmit={this.onSubmit}
          />          
        </div>
        <CategoryList />
      </div>
    )
  };
}

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category))
});

export default connect(undefined, mapDispatchToProps)(CategoriesPage);