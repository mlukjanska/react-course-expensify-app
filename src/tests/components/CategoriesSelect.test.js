import React from 'react';
import { shallow } from 'enzyme';
import { CategoriesSelect } from '../../components/CategoriesSelect';
import categories from '../fixtures/categories'

let onCategoryChange, wrapper;

beforeEach(() => {
    onCategoryChange = jest.fn();
});

test('should render CategoriesSelect correctly without selected category', () => {
    const categorySelected = '';
    wrapper = shallow(
        <CategoriesSelect 
            categories={categories}
            categorySelected={categorySelected} 
            onCategoryChange={onCategoryChange} 
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('should render CategoriesSelect correctly with selected category', () => {
    const categorySelected = categories[0].value;
    wrapper = shallow(
        <CategoriesSelect 
            categories={categories}
            categorySelected={categorySelected} 
            onCategoryChange={onCategoryChange} 
        />
    );
    expect(wrapper).toMatchSnapshot();
});