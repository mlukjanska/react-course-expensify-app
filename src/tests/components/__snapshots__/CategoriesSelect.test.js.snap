// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CategoriesSelect correctly with selected category 1`] = `
<div>
  <StateManager
    className="react-select"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isClearable={true}
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": "1",
          "label": "Food",
          "value": "food",
        },
        Object {
          "id": "2",
          "label": "Clothes",
          "value": "clothes",
        },
        Object {
          "id": "3",
          "label": "House utilities",
          "value": "house-utilities",
        },
        Object {
          "id": "4",
          "label": "Bills",
          "value": "bills",
        },
      ]
    }
    placeholder="Select category..."
  />
</div>
`;

exports[`should render CategoriesSelect correctly without selected category 1`] = `
<div>
  <StateManager
    className="react-select"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isClearable={true}
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": "1",
          "label": "Food",
          "value": "food",
        },
        Object {
          "id": "2",
          "label": "Clothes",
          "value": "clothes",
        },
        Object {
          "id": "3",
          "label": "House utilities",
          "value": "house-utilities",
        },
        Object {
          "id": "4",
          "label": "Bills",
          "value": "bills",
        },
      ]
    }
    placeholder="Select category..."
    value=""
  />
</div>
`;
